### DISCUSSION TITLE
Color Inversion

### DISCUSSION THREAD

**Discussion by AppleSheeple** [YaLTeR/niri#163] [Ideas]
Searching didn't return any results, so I thought I would create this feature request me and others can subscribe to.

Color inversion, preferably customizable  picom/compton-style color inversion, is the primary reason why I'm still on X11.

(Anyone considering mentioning gamma control shall slap themselves before doing so.)

**Comment by YaLTeR** [comment]
Hmm, what do you mean? Not sure I ever came across a feature like this

**Reply by AppleSheeple** [reply]
It's just a general "color inversion" feature request.

If you mean the `picom` part, with `picom`, you can run a custom GLSL shader that inverts colors the way you like.

For example:

```
uniform bool invert_color;
uniform float opacity;
uniform sampler2D tex;

void main() {
  vec4 c = texture2D(tex, gl_TexCoord[0].st);
  if (invert_color) {
    // white bias
    float wb = 0.03;
    c = vec4(wb +  c.a + c.r*0.18 - c.g*0.57 - c.b*0.57,
             wb +  c.a + c.g*0.18 - c.r*0.57 - c.b*0.57,
             wb +  c.a + c.b*0.18 - c.r*0.57 - c.g*0.57,
             wb + c.a);
  }
  c *= opacity;
  gl_FragColor = c;
}
```

which looks like this:

![invert](https://github.com/YaLTeR/niri/assets/68065381/dda96144-af53-494e-a6ca-454d0e260a10)

**Reply by YaLTeR** [reply]
I think running a custom shader shouldn't be too difficult to do; it does kill direct scanout though so there's some performance drawback.

**Reply by AppleSheeple** [reply]
> so there's some performance drawback.

That's fine by me. I don't think anyone watches videos\* or plays games with color inversion on.

\* Exceptions exist like presentations with light background. But if this is indeed implementable, I don't think the performance hit will be big enough to prevent that anyway.

**Reply by coreymwamba** [reply]
I came here to advocate for something like this. I used picom to set the screen to greyscale, or adjust the brightness. The syntax I used to use was `picom -b --legacy-backends --backend glx --glx-no-stencil --xrender-sync-fence --glx-fshader-win "$(cat $HOME/.config/glsl/greyscale.glsl)" `, using `cat` because entering the shader in a line was clunky, and I preferred storing the shaders in a file. 

I was thinking of maybe something like `<key-binding> { apply-screen-shader "file.glsl"; }`

Could niri potentially use filenames to implement custom shaders, rather than entering the text in the config file?
